src/main/java/com/amazon/opendistroforelasticsearch/sql/utils/SQLFunctions.java

```diff
@@ -620,6 +665,69 @@ private String upper(String property, String culture) {
         return property + ".toUpperCase(Locale.forLanguageTag(\"" + culture + "\"))";
     }
 
+    private Tuple<String, String> length(SQLExpr field, String valueName) {
+        String name = nextId("length");
+        if (Strings.isNullOrEmpty(valueName)) {
+            return new Tuple<>(name, def(name, getPropertyOrStringValue(field) + ".length()"));
+        } else {
+            return new Tuple<>(name, getPropertyOrValue(field) + ";" + def(name, valueName + ".length()"));
+        }
+    }
+
+    private Tuple<String, String> replace(SQLExpr field, String target, String replacement, String valueName) {
+        String name = nextId("replace");
+        if (Strings.isNullOrEmpty(valueName)) {
+            return new Tuple<>(name, def(name, getPropertyOrStringValue(field)
+                    + ".replace(" + target + "," + replacement + ")"));
+        } else {
+            return new Tuple<>(name, getPropertyOrStringValue(field) + ";"
+                    + def(name, valueName + ".replace(\" + target + \",\" + replacement + \")"));
+        }
+    }
+
+    // es behavior: both 'start' and return value are 1-index; return 0 if pattern does not exist;
+    // support out-of-bound index
+    private Tuple<String, String> locate(String pattern, SQLExpr source, int start) {
+        String name = nextId("locate");
+        String docSource = getPropertyOrStringValue(source);
+        start = start < 1 ? 0 : start - 1;
+        return new Tuple<>(name, StringUtils.format(
+                "int count=%d;int res=-1;" + "while(count< %s.length() - %s.length()) {"
+                        + "if (%s == %s.substring(count, count+%s.length())) {"
+                        + "res = count;break;} "
+                        + "else {count++;}}"
+                        + def(name, "res+1"),
```

Comment:
If I understand right, this `locate` function is essentially `indexOf` in Java/Painless. Could we make use of `indexOf(String, int)` API directly?

Reply:
I've updated them to indexOf(), thanks!